### PR TITLE
[HttpKernel] Add `skip_response_headers` to the `HttpCache` options

### PR DESCRIPTION
**Pull request by Toflar**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | https://github.com/symfony/symfony/issues/49257
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php [hunk 1]
```diff
@@ -591,10 +595,19 @@ protected function lock(Request $request, Response $entry): bool
     protected function store(Request $request, Response $response)
     {
         try {
-            $this->store->write($request, $response);
+            $restoreHeaders = [];
+            foreach ($this->options['skip_response_headers'] as $header) {
+                $restoreHeaders[$header] = $response->headers->get($header);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php]: missing "if has($header)" around?

File: src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php [hunk 2]
```diff
@@ -591,10 +595,19 @@ protected function lock(Request $request, Response $entry): bool
...
+                $response->headers->remove($header);
+            }
 
+            $this->store->write($request, $response);
             $this->record($request, 'store');
 
+            foreach ($restoreHeaders as $header => $values) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php]: should be in a try/finally

File: src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php [hunk 3]
```diff
@@ -1740,6 +1740,42 @@ public function testResponsesThatMustNotBeUsedStaleIfError($responseHeaders, $sl
...
+                $this->isInstanceOf(Request::class),
+                $this->callback(function (Response $response) {
+                    $this->assertFalse($response->headers->has('Set-Cookie'));
+                    $this->assertFalse($response->headers->has('Another-One-To-Skip'));
+                    $this->assertTrue($response->headers->has('Cache-Control'));
+                    $this->assertTrue($response->headers->has('Another-One-To-Keep'));
+
+                    return true;
+                }
+                ));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php]: should be unindented a bit

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Tests/HttpCache/HttpCacheTest.php]: last bracket should be on its own line

**Review by nicolas-grekas — COMMENTED** [review]
LGTM thanks, I just had some minor comments.

File: src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php (hunk 1 above)
Comment (Toflar) [reply src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php]: Fixed and adjusted tests.

File: src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php [hunk 4]
```diff
@@ -591,8 +595,17 @@ protected function lock(Request $request, Response $entry): bool
...
     {
         try {
-            $this->store->write($request, $response);
+            $restoreHeaders = [];
+            foreach ($this->options['skip_response_headers'] as $header) {
+                if (!$response->headers->has($header)) {
+                    continue;
+                }
 
+                $restoreHeaders[$header] = $response->headers->get($header);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php]: shouldn't we use `all($header)` here ?

Comment (Toflar) [reply src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php]: Of course, I thought `get` would already return an array. I guess that used to be like that back in the days :D

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php [hunk 5]
```diff
@@ -60,6 +60,9 @@ class HttpCache implements HttpKernelInterface, TerminableInterface
      *                            on responses that don't explicitly state whether the response is
      *                            public or private via a Cache-Control directive. (default: Authorization and Cookie)
      *
+     *   * skip_response_headers  Set of response headers that are never cached even if a response is cacheable (public).
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php:63]: I think `skipped_response_headers` might be a better name for the option

**Review by maxhelias — APPROVED** [review]

File: src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php:63]: let's keep it as is, it also makes sense, is shorter and prevents reindenting all the docblock :P

Comment (Toflar) [reply src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php:63]: The docblock was the reason for me to keep it as short as it is now 🙈

**Comment by nicolas-grekas** [comment]
Thank you @Toflar.

**Pull request merged**